### PR TITLE
Implement #[alloc_error_handler]

### PR DESCRIPTION
**Pull request by SimonSapin**
This to-be-stable attribute is equivalent to `#[lang = "oom"]`. It is required when using the alloc crate without the std crate. It is called by `handle_alloc_error`, which is in turned called by "infallible" allocations APIs such as `Vec::push`.

Tracking issue: https://github.com/rust-lang/rust/issues/51540

### CODE REVIEW

**Comment by SimonSapin** [comment]
CC @alexcrichton, @japaric, @glandium, @Amanieu

**Comment by glandium** [comment]
Random thought. If you can't use no_std + alloc without global_allocator *and* alloc_error_handler, why not put the allocation error handling in the GlobalAlloc trait, like it (kind of) used to be (because that was on the Alloc trait before GlobalAlloc existed). That would even allow a default to exist!

**Comment by alexcrichton** [comment]
@glandium the original thinking is that they're separable concerns especially at the libstd layer. Right now this implementation of handling OOM is different from libstd's allocator implementation in that it doesn't provide a fallback automatically if you haven't otherwise provided one. For libstd it makes sense to define how to allocate without defining how to implement OOM, and the same could perhaps one day also be true of `alloc` with enough work going into this attribute.

File: src/libstd/alloc.rs [hunk 1]
```diff
@@ -125,7 +125,8 @@ fn default_alloc_error_hook(layout: Layout) {
 
 #[cfg(not(test))]
 #[doc(hidden)]
-#[lang = "oom"]
+#[cfg_attr(stage0, lang = "oom")]
+#[cfg_attr(not(stage0), alloc_error_handler)]
 #[unstable(feature = "alloc_internals", issue = "0")]
 pub extern fn rust_oom(layout: Layout) -> ! {
```
Comment (ollie27) [inline src/libstd/alloc.rs]: What abi is this function supposed to have? Here it's `"C"` but elsewhere it seems to be `"Rust"`.

Comment (SimonSapin) [reply src/libstd/alloc.rs]: Good point, I’ve removed `extern` to make it Rust ABI.

**Comment by SimonSapin** [comment]
Failure above was `curl error: Couldn't resolve host 'github.com'`

@bors r=alexcrichton

File: src/libstd/alloc.rs (hunk 1 above)
Comment (ollie27) [reply src/libstd/alloc.rs]: Shouldn't the abi be checked as part of the signature check?

**Comment by alexcrichton** [comment]
Er sorry, was cleaning out branches on the `rust-lang` repo and accidentally closed this, can the PR be resent from your own repo @SimonSapin?

**Pull request closed without merging**